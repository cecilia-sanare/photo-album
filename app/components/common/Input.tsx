import React from 'react';
import * as styles from './Input.module.scss';
import classNames from 'classnames';
import { useCachedState } from '../../hooks/use-cached-state';

type InputProps = {
  className?: string;
  value?: string;
  onChange?: (value: string) => void;
} & React.DetailedHTMLProps<React.InputHTMLAttributes<HTMLInputElement>, HTMLInputElement>;

export function Input({
  className,
  value,
  onChange,
  ...extraProps
}: InputProps) {
  const [internalValue, setInternalValue] = useCachedState<string | null>(() => value, [value]);

  return (
    <input
      className={classNames(
        styles.input,
        className
      )}
      value={internalValue}
      onChange={(event) => {
        setInternalValue(event.target.value);
        onChange?.(event.target.value);
      }}
      {...extraProps}
    />
  );
}
