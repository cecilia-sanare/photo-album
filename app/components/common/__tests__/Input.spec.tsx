import { fireEvent, render } from '@testing-library/react';
import { Input } from '../Input';
import Chance from 'chance';

const chance = new Chance();

describe('<Input />', () => {
  it('should render the input', () => {
    const component = render(<Input />);

    expect(component.baseElement).toMatchSnapshot();
  });

  describe('prop(value)', () => {
    it('should support providing a value', () => {
      const expectedValue = chance.string();

      const component = render(<Input value={expectedValue} />);

      expect(component.getByDisplayValue(expectedValue)).toBeTruthy();
    });
  });

  describe('prop(onChange)', () => {
    it('should support providing an onChange callback', async () => {
      const onChange = jest.fn();
      const expectedValue = chance.string();

      const component = render(<Input onChange={onChange} />);
      const input = component.getByTestId('input');

      expect(onChange).toHaveBeenCalledTimes(0);

      await fireEvent.change(input, {target: {value: expectedValue}})

      expect(onChange).toHaveBeenCalledTimes(1);
      expect(onChange).toHaveBeenCalledWith(expectedValue);
    });

    it('should support not providing an onChange callback', async () => {
      const expectedValue = chance.string();

      const component = render(<Input />);
      const input = component.getByTestId('input');

      expect(component.queryByDisplayValue(expectedValue)).toBeFalsy();

      await fireEvent.change(input, {target: {value: expectedValue}})

      expect(component.getByDisplayValue(expectedValue)).toBeTruthy();
    });
  });
});
