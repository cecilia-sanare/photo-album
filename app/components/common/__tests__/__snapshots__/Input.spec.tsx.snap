// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Input /> should render the input 1`] = `
<body>
  <div>
    <input
      class="input"
      data-testid="input"
      value=""
    />
  </div>
</body>
`;
